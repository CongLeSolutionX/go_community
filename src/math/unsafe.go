// Copyright 2009 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package math

import "unsafe"

// Float32bits returns the IEEE 754 binary representation of f.
// the most significant bit of the return value will contain the sign bit,
// the next 8 most significant bits the exponent,
// and the 23 least significant bits the significand or mantissa.
func Float32bits(f float32) uint32 { return *(*uint32)(unsafe.Pointer(&f)) }

// Float32frombits returns the floating point number corresponding
// to the IEEE 754 binary representation b.
// The most significant bit of b should contain the sign bit,
// the next 8 most significant bits the exponent,
// and the 23 least significant bits the significand or mantissa.
func Float32frombits(b uint32) float32 { return *(*float32)(unsafe.Pointer(&b)) }

// Float64bits returns the IEEE 754 binary representation of f.
// the most significant bit of the return value will contain the sign bit,
// the next 12 most significant bits the exponent,
// and the 51 least significant bits the significand or mantissa.
func Float64bits(f float64) uint64 { return *(*uint64)(unsafe.Pointer(&f)) }

// Float64frombits returns the floating point number corresponding
// the IEEE 754 binary representation b.
// The most significant bit of b should contain the sign bit,
// the next 12 most significant bits the exponent,
// and the 51 least significant bits the significand or mantissa.
func Float64frombits(b uint64) float64 { return *(*float64)(unsafe.Pointer(&b)) }
