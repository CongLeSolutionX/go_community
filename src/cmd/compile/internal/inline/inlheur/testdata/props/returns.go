// Copyright 2023 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package returns1

import "unsafe"

type Bar struct {
	x int
	y string
}

func (b *Bar) Plark() {
}

type Q int

func (q *Q) Plark() {
}

type Itf interface {
	Plark()
}

func T_simple_allocmem() *Bar {
	return &Bar{}
}

func T_allocmem_two_returns(x int) *Bar {
	// multiple returns
	if x < 0 {
		return new(Bar)
	} else {
		return &Bar{x: 2}
	}
}

func T_allocmem_three_returns(x int) []*Bar {
	// more multiple returns
	switch x {
	case 10, 11, 12:
		return make([]*Bar, 10)
	case 13:
		fallthrough
	case 15:
		return []*Bar{&Bar{x: 15}}
	}
	return make([]*Bar, 0, 10)
}

func T_return_nil() *Bar {
	// simple case: no alloc
	return nil
}

func T_multi_return_nil(x, y bool) *Bar {
	if x && y {
		return nil
	}
	return nil
}

func T_multi_return_nil_anomoly(x, y bool) Itf {
	if x && y {
		var qnil *Q
		return qnil
	}
	var barnil *Bar
	return barnil
}

func T_multi_return_some_nil(x, y bool) *Bar {
	if x && y {
		return nil
	} else {
		return &GB
	}
}

var GB Bar

func T_mixed_returns(x int) *Bar {
	// mix of alloc and non-alloc
	if x < 0 {
		return new(Bar)
	} else {
		return &GB
	}
}

func T_mixed_returns_slice(x int) []*Bar {
	// mix of alloc and non-alloc
	switch x {
	case 10, 11, 12:
		return make([]*Bar, 10)
	case 13:
		fallthrough
	case 15:
		return []*Bar{&Bar{x: 15}}
	}
	ba := [...]*Bar{&GB, &GB}
	return ba[:]
}

func T_maps_and_channels(x int, b bool) (bool, map[int]int, chan bool, unsafe.Pointer) {
	// maps and channels
	return b, make(map[int]int), make(chan bool), nil
}

func T_assignment_to_named_returns(x int) (r1 *uint64, r2 *uint64) {
	// assignments to named returns and then "return" not supported
	r1 = new(uint64)
	if x < 1 {
		*r1 = 2
	}
	r2 = new(uint64)
	return
}

func T_named_returns_but_return_explicit_values(x int) (r1 *uint64, r2 *uint64) {
	// named returns ok if we don't use them by name
	rx1 := new(uint64)
	if x < 1 {
		*rx1 = 2
	}
	rx2 := new(uint64)
	return rx1, rx2
}

func T_return_concrete_type_to_itf(x, y int) Itf {
	return &Bar{}
}

func T_return_concrete_type_to_itfwith_copy(x, y int) Itf {
	b := &Bar{}
	println("whee")
	return b
}

func T_return_concrete_type_to_itf_mixed(x, y int) Itf {
	if x < y {
		b := &Bar{}
		return b
	}
	return nil
}
