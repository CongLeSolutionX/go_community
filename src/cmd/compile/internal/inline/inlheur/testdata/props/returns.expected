// DO NOT EDIT (use 'go test -update-expected' instead))
(*Bar).Plark
===
{
	"Flags": 0,
	"RecvrParamFlags": null,
	"ReturnFlags": []
}
=-=-=-
(*Q).Plark
===
{
	"Flags": 0,
	"RecvrParamFlags": null,
	"ReturnFlags": []
}
=-=-=-
T_simple_allocmem
// ReturnFlags:
//   0: ReturnIsAllocatedMem
===
{
	"Flags": 0,
	"RecvrParamFlags": null,
	"ReturnFlags": [
		2
	]
}
=-=-=-
T_allocmem_two_returns
// ReturnFlags:
//   0: ReturnIsAllocatedMem
===
{
	"Flags": 0,
	"RecvrParamFlags": null,
	"ReturnFlags": [
		2
	]
}
=-=-=-
T_allocmem_three_returns
// ReturnFlags:
//   0: ReturnIsAllocatedMem
===
{
	"Flags": 0,
	"RecvrParamFlags": null,
	"ReturnFlags": [
		2
	]
}
=-=-=-
T_return_nil
// ReturnFlags:
//   0: ReturnAlwaysSameConstant
===
{
	"Flags": 0,
	"RecvrParamFlags": null,
	"ReturnFlags": [
		8
	]
}
=-=-=-
T_multi_return_nil
// ReturnFlags:
//   0: ReturnAlwaysSameConstant
===
{
	"Flags": 0,
	"RecvrParamFlags": null,
	"ReturnFlags": [
		8
	]
}
=-=-=-
T_multi_return_nil_anomoly
===
{
	"Flags": 0,
	"RecvrParamFlags": null,
	"ReturnFlags": [
		0
	]
}
=-=-=-
T_multi_return_some_nil
===
{
	"Flags": 0,
	"RecvrParamFlags": null,
	"ReturnFlags": [
		0
	]
}
=-=-=-
T_mixed_returns
===
{
	"Flags": 0,
	"RecvrParamFlags": null,
	"ReturnFlags": [
		0
	]
}
=-=-=-
T_mixed_returns_slice
===
{
	"Flags": 0,
	"RecvrParamFlags": null,
	"ReturnFlags": [
		0
	]
}
=-=-=-
T_maps_and_channels
// ReturnFlags:
//   0: ReturnNoInfo
//   1: ReturnNoInfo
//   2: ReturnNoInfo
//   3: ReturnAlwaysSameConstant
===
{
	"Flags": 0,
	"RecvrParamFlags": null,
	"ReturnFlags": [
		0,
		0,
		0,
		8
	]
}
=-=-=-
T_assignment_to_named_returns
===
{
	"Flags": 0,
	"RecvrParamFlags": null,
	"ReturnFlags": [
		0,
		0
	]
}
=-=-=-
T_named_returns_but_return_explicit_values
// ReturnFlags:
//   0: ReturnIsAllocatedMem
//   1: ReturnIsAllocatedMem
===
{
	"Flags": 0,
	"RecvrParamFlags": null,
	"ReturnFlags": [
		2,
		2
	]
}
=-=-=-
T_return_concrete_type_to_itf
// ReturnFlags:
//   0: ReturnIsConcreteTypeConvertedToInterface
===
{
	"Flags": 0,
	"RecvrParamFlags": [
		0,
		0
	],
	"ReturnFlags": [
		4
	]
}
=-=-=-
T_return_concrete_type_to_itfwith_copy
// ReturnFlags:
//   0: ReturnIsConcreteTypeConvertedToInterface
===
{
	"Flags": 0,
	"RecvrParamFlags": [
		0,
		0
	],
	"ReturnFlags": [
		4
	]
}
=-=-=-
T_return_concrete_type_to_itf_mixed
===
{
	"Flags": 0,
	"RecvrParamFlags": [
		0,
		0
	],
	"ReturnFlags": [
		0
	]
}
=-=-=-
.eq.command-line-arguments.Bar
===
{
	"Flags": 0,
	"RecvrParamFlags": null,
	"ReturnFlags": [
		0
	]
}
=-=-=-
init
===
{
	"Flags": 0,
	"RecvrParamFlags": null,
	"ReturnFlags": []
}
=-=-=-
